[ALLUXIO-2639] - List all pinned files in Alluxio shell

ShailShah opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-2639

AmplabJenkins commented:
Can one of the admins verify this patch?

uronce-cc commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -173,8 +193,9 @@ public String getUsage() {
-        + " Specify -R to display files and directories recursively."
-        + " Specify -d to list directories as plain files."
-        + " Specify -f to force loading files in the directory.";
+            + " Specify -R to display files and directories recursively."
```
please fix the indentation

ShailShah replied:
@uronce-cc Have fixed the coding style. Could you please review it.

uronce-cc commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -128,20 +137,31 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
   private List<URIStatus> listStatusSortedByIncreasingCreationTime(AlluxioURI path,
-      ListStatusOptions options)
-      throws AlluxioException, IOException {
+                                                                   ListStatusOptions options)
```
please fix the style

uronce-cc commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -44,15 +44,15 @@
-   * @param userName user name
-   * @param groupName group name
-   * @param size size of the file in bytes
+   * @param acl          whether security is enabled
```
please keep the original style, no extra spaces are needed to be aligned, apply to the whole PR.

uronce-cc commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -99,26 +99,33 @@ protected Options getOptions() {
+        }
+      }
       System.out.print(formatLsString(SecurityUtils.isSecurityEnabled(), status.isFolder(),
-          FormatUtils.formatMode((short) status.getMode(), status.isFolder()), status.getOwner(),
```
keep the original style

uronce-cc commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -175,6 +192,7 @@ public String getDescription() {
         + " Specify -d to list directories as plain files."
-        + " Specify -f to force loading files in the directory.";
+        + " Specify -f to force loading files in the directory."
+        + " Specify -p to list all the pinned files in the specified path";
```
remove `in the specified path` since the semantic is already expressed above

uronce-cc commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -99,26 +99,33 @@ protected Options getOptions() {
    */
-  private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, boolean dirAsFile)
-      throws AlluxioException, IOException {
+  private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, boolean dirAsFile,
```
it will be more maintainable to move `recursive`, `forceLoadMetadata`, `dirAsFile`, `pinnedMetadata` to `ListStatusOptions`. also s/pinnedMetadata/pinned, it's simpler and clear.

jsimsa replied:
@uronce-cc I agree that the use of options might help here, but the `alluxio.client.file.options.ListStatusOptions` is meant to be use for options passed to Alluxio servers through `FileSystem` interface invocations ... it makes no sense to be introducing options that have no effect on the `FileSystem` interface invocations

uronce-cc commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -128,20 +135,30 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
-          100 == status.getInMemoryPercentage(), status.getPath()));
+      if (pinnedMetadata) {
+        if (status.isPinned()) {
+          System.out.print(formatLsString(SecurityUtils.isSecurityEnabled(), status.isFolder(),
```
this block is the same as the `else` block, and appears in other places, could you consider abstracting it out as a private method.

ShailShah commented:
Have modified and included all the options in ListStatusOptions. But, I haven't modified ListStatusTOptions.

yupeng9 commented:
@uronce-cc do you have more comments?

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @return raw size option. It specifies whether to print raw sizes
+   */
+  public boolean ismRawSize() {
```
`isRawSize`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
   }
 
+  /**
+   * @return raw size option. It specifies whether to print raw sizes
```
nit: please end sentences with a `.`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  }
+
+  /**
+   * @param rawSize Raw Size Option
```
replace "Raw Size Option" with "the value for the raw size flag to use"

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @param rawSize Raw Size Option
+   */
+  public void setmRawSize(boolean rawSize) {
```
`setRawSize`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  }
+
+  /**
+   * @return the recursive option. It specifies whether the path has to be traversed recursively
```
nit: please end sentences with a `.`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @return the recursive option. It specifies whether the path has to be traversed recursively
+   */
+  public boolean ismRecursive() {
```
`isRecursive`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  }
+
+  /**
+   * @param recursive Recursive option. It specifies whether to display directories recursively
```
replace "Recursive option" with "the value of the recursive flag to use"

jsimsa replied:
nit: please end sentences with a `.`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  }
+
+  /**
+   * @return Force Load Metadata Option. It specifies whether to force loading files in the
```
replace "Force Load Metadata Option" with "the force load metadata option"

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @param recursive Recursive option. It specifies whether to display directories recursively
+   */
+  public void setmRecursive(boolean recursive) {
```
`setRecursive`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+   * @return Force Load Metadata Option. It specifies whether to force loading files in the
+   *         directory.
+   */
+  public boolean ismForceLoadMetaData() {
```
`isForceLoadMetadata`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  }
+
+  /**
+   * @param forceLoadMetaData Option to force loading files in the directory
```
`@param forceLoadMetadata the force load metadata option to use. It specifies whether to force loading files in the directory.`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @param forceLoadMetaData Option to force loading files in the directory
+   */
+  public void setmForceLoadMetaData(boolean forceLoadMetaData) {
```
`setForceLoadMetadata(boolean forceLoadMetadata)`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  }
+
+  /**
+   * @return Directory As File Option. It specifies whether to list directories as plain files
```
what is this option for? can you describe your use case?

jsimsa replied:
`the directory as file option. It specifies whether to list directories as plain files.`

ShailShah replied:
@jsimsa This option (the directory as file option) was available in the LsCommand before I made any changes. I don't have any use case for this option.

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @return Directory As File Option. It specifies whether to list directories as plain files
+   */
+  public boolean ismDirAsFile() {
```
`isDirAsFile`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  }
+
+  /**
+   * @param dirAsFile Option to list directories as plain files
```
`@param the value of the the directory as file flag to use. It specifies whether to list directories as plain files.`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @param dirAsFile Option to list directories as plain files
+   */
+  public void setmDirAsFile(boolean dirAsFile) {
```
`setDirAsFile`

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
+  /**
+   * @return IsPinned Option. It specifies whether to list all the pinned files
+   */
+  public boolean ismPinned() {
```
change the name of this flag to `pinnedOnly` and update the setter and getter following the style suggested in my previous comments

jsimsa commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -170,12 +172,21 @@ public int compare(URIStatus status1, URIStatus status2) {
   @Override
   public String getUsage() {
-    return "ls [-R|-d|-f|-raw] <path>";
+    return "ls [-R|-d|-f|-raw|-p] <path>";
```
keep options sorted alphabetically

ShailShah commented:
@jsimsa I have made the requested changes. PTAL. I got confused about the usage of ListStatusOptions class. I have removed the options from the ListStatusOptions class and reverted it back to the original state. I have added the `-p` option in the LsCommand itself instead of creating the new class.

jsimsa commented on core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -15,8 +15,6 @@
 import alluxio.thrift.ListStatusTOptions;
 import alluxio.wire.LoadMetadataType;
 
-import com.fasterxml.jackson.annotation.JsonInclude;
```
please undo these changes

jsimsa commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -117,15 +125,17 @@ protected Options getOptions() {
    */
   private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, boolean dirAsFile,
-                  boolean rawSize)
+                  boolean rawSize, boolean pinned)
```
rename `pinned` to `pinnedOnly`

jsimsa commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -135,21 +145,22 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
-          FormatUtils.formatMode((short) status.getMode(), status.isFolder()), status.getOwner(),
-          status.getGroup(), status.getLength(), status.getCreationTimeMs(),
-          100 == status.getInMemoryPercentage(), status.getPath()));
+      if (pinned) {
```
replace the whole if-then-else block with

```
if (!pinnedOnly || status.isPinned()) {
  printLsString(status, rawSize);
}
```

jsimsa commented on shell/src/main/java/alluxio/shell/command/LsCommand.java:
```diff
@@ -184,6 +196,7 @@ public String getDescription() {
         + " Specify -d to list directories as plain files."
         + " Specify -f to force loading files in the directory."
-        + " Specify -raw to print raw sizes.";
+        + " Specify -raw to print raw sizes."
```
keep options sorted alphabetically

ShailShah commented:
@jsimsa Made the required changes, PTAL.

ShailShah commented:
@jsimsa Have added the test for `-p` functionality. PTAL.

jsimsa commented:
Jenkins, test this please

jsimsa commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13998/
Test PASSed.

jsimsa commented:
@ShailShah thank you for your contribution
